presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/LogicalCteOptimizer.java

```diff
@@ -0,0 +1,221 @@
+        {
+            return new ApplyNode(node.getSourceLocation(),
+                    idAllocator.getNextId(),
+                    context.rewrite(node.getInput(),
+                            context.get()),
+                    context.rewrite(node.getSubquery(),
+                            context.get()),
+                    node.getSubqueryAssignments(),
+                    node.getCorrelation(),
+                    node.getOriginSubqueryError(),
+                    node.getMayParticipateInAntiJoin());
+        }}
+
+    public class CteTransformerContext
+    {
+        private static final int VISITED = 1;
+
+        private static final int EXPLORING = 2;
+        public Map<String, CteProducerNode> cteProducerMap;
+
+        // a -> b indicates that b needs to be processed before a
+        public Map<String, List<String>> cteDependencyGraph;
+
+        public String activeCte;
+
+        public CteTransformerContext()
+        {
+            cteProducerMap = new HashMap<>();
+            cteDependencyGraph = new HashMap<>();
+            activeCte = "";
+        }
+
+        public Map<String, CteProducerNode> getCteProducerMap()
+        {
+            return cteProducerMap;
+        }
+
+        public void addProducer(String cteName, CteProducerNode cteProducer)
+        {
+            cteProducerMap.putIfAbsent(cteName, cteProducer);
+        }
+
+        public void setActiveCte(String activeCte)
+        {
+            this.activeCte = activeCte;
+        }
+
+        public String getActiveCte()
+        {
+            return activeCte;
+        }
+
+        public void clearActiveCte()
+        {
+            this.activeCte = "";
+        }
+
+        public void addDependency(String currentCte)
+        {
+            if (!getActiveCte().equals("")) {
+                List<String> adjacencyList = cteDependencyGraph.getOrDefault(activeCte, new ArrayList<>());
+                adjacencyList.add(currentCte);
+                cteDependencyGraph.put(activeCte, adjacencyList);
+            }
+            cteDependencyGraph.putIfAbsent(currentCte, new ArrayList<>());
+        }
+
+        public List<PlanNode> getTopologicalOrdering()
+        {
+            Map<String, Integer> visited = new HashMap<>();
+            List<String> topSortList = new ArrayList<>();
+            for (String node : cteDependencyGraph.keySet()) {
+                if (!visited.containsKey(node)) {
+                    dfs(node, topSortList, visited);
+                }
+            }
+            // DFS will generate reverse order
+            Collections.reverse(topSortList);
+            List<PlanNode> topSortedCteProducerList = new ArrayList<>();
+            for (String cteName : topSortList) {
+                topSortedCteProducerList.add(cteProducerMap.get(cteName));
+            }
+            return topSortedCteProducerList;
+        }
+
+        private void dfs(String node, List<String> topSortList, Map<String, Integer> visitedMap)
```

Comment:
You'll find examples in the codebase that we use the Guava Traverser utility to do a DFS, wondering if we can refactor this to use it?  May reduce the LOCs.

Reply:
I had initially used the guava implementation but detecting cycle there is nontrivial and unclean. I think LOC is an anti goal

Reply:
Actually I should probably check recursive ctes and add test cases if possible

Reply:
Recursive ctes are not supported in Presto

Reply:
Actually I found a way to change the graph to use MutableGraph (marked unstable) and the complexity reduces significantly, please rereview

Reply:
But we would need the cycle detection logic when we do the cost based replacement to filter out unwanted plans so i might use the old implementation there anyways